pkg/bgpv1/agent/controller.go

```diff
@@ -174,35 +195,43 @@ func configureForClusterPoolIPAM(factory externalversions.SharedInformerFactory,
 // The constructor requires an implementation of BGPRouterManager to be provided.
 // This implementation defines which BGP backend will be used (GoBGP, FRR, Bird, etc...)
 // NOTE: only GoBGP currently implemented.
-//
-// Cancel the provided CTX to stop the Controller.
-func NewController(ctx context.Context, clientset client.Clientset, rtMgr BGPRouterManager, opts ...ControllerOpt) (*Controller, error) {
+func NewController(params ControllerParams) (*Controller, error) {
+	// If the BGP control plane is disabled, just return nil. This way the hive dependency graph is always static
+	// regardless of config. The lifecycle has not been appended so no work will be done.
+	if !params.DaemonConfig.BGPControlPlaneEnabled() {
+		return nil, nil
+	}
+
+	log.Info("Initializing BGP Control Plane")
+
 	var (
 		// signaler used to trigger Controller reconciliation.
 		sig = NewSignaler()
-		// stop channel used to cancel informers.
-		stop = make(chan struct{}, 1)
 		// we'll use to list and watch BGPPeeringPolicies
-		factory = externalversions.NewSharedInformerFactory(clientset, 0)
-		// an abtraction over obtaining Node Spec information depending on
-		// IPAM configuration.
-		nodeSpecer nodeSpecer
+		factory = externalversions.NewSharedInformerFactory(params.Clientset, 0)
 	)
 
+	c := Controller{
+		Sig:    sig,
+		BGPMgr: params.RouteMgr,
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
```

Comment:
Is there any reason not to do this in the `Start` hook? https://github.com/cilium/cilium/pull/22183/files#diff-663c8aba9492de183ebe045c8bf425d6a6a645970cd2a201a7073e9b21ffe8adR272

This just feels a bit better as you're creating the context where the runtime will set the controller into action.

Reply:
Adopted this pattern due to feedback from other reviewers on #21764. I can see an argument for doing it both ways, will move it to `Start()`.

Reply:
Consider whether [workerpool](https://github.com/cilium/workerpool) would be a good fit here. E.g. instead of `go c.Run()` etc. you'd do `c.workerpool.Submit("run", c.Run)` and on Stop() you'd `c.workerpool.Close()` which cancels the context passed to each worker and waits for them to finish.

Reply:
Yes, I think it is, given we have 3 seperate go routines in Start() as well